src/include/libpmemobj.h

```diff
@@ -84,6 +99,553 @@ void pmemobj_set_funcs(
 		void *(*malloc_func)(size_t size),
 		void (*free_func)(void *ptr));
 
+/*
+ * Locking...
+ */
+#define	PMEMOBJ_CL_ALIGNMENT 64 /* cache line alignment for performance */
+
+typedef union padded_pmemmutex {
+	char padding[PMEMOBJ_CL_ALIGNMENT];
+	struct {
+		uint64_t runid;
+		pthread_mutex_t mutex;
+	} pmemmutex;
+} PMEMmutex;
+
+typedef union padded_pmemrwlock {
+	char padding[PMEMOBJ_CL_ALIGNMENT];
+	struct {
+		uint64_t runid;
+		pthread_rwlock_t rwlock;
+	} pmemrwlock;
+} PMEMrwlock;
+
+typedef union padded_pmemcond {
+	char padding[PMEMOBJ_CL_ALIGNMENT];
+	struct {
+		uint64_t runid;
+		pthread_cond_t cond;
+	} pmemcond;
+} PMEMcond;
+
+int pmemobj_mutex_zero(PMEMmutex *mutexp);
+int pmemobj_mutex_lock(PMEMobjpool *pop, PMEMmutex *mutexp);
+int pmemobj_mutex_trylock(PMEMobjpool *pop, PMEMmutex *mutexp);
+int pmemobj_mutex_unlock(PMEMobjpool *pop, PMEMmutex *mutexp);
+
+int pmemobj_rwlock_zero(PMEMrwlock *rwlockp);
+int pmemobj_rwlock_rdlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_wrlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_timedrdlock(PMEMobjpool *pop,
+	PMEMrwlock *restrict rwlockp,
+	const struct timespec *restrict abs_timeout);
+int pmemobj_rwlock_timedwrlock(PMEMobjpool *pop,
+	PMEMrwlock *restrict rwlockp,
+	const struct timespec *restrict abs_timeout);
+int pmemobj_rwlock_tryrdlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_trywrlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_unlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+
+int pmemobj_cond_zero(PMEMcond *condp);
+int pmemobj_cond_broadcast(PMEMobjpool *pop, PMEMcond *condp);
+int pmemobj_cond_signal(PMEMobjpool *pop, PMEMcond *condp);
+int pmemobj_cond_timedwait(PMEMobjpool *pop, PMEMcond *restrict condp,
+	PMEMmutex *restrict mutexp, const struct timespec *restrict abstime);
+int pmemobj_cond_wait(PMEMobjpool *pop, PMEMcond *condp,
+	PMEMmutex *restrict mutexp);
+
+/* find a heap offset of pool identified by uuid_lo */
+uint64_t pmemobj_heap_off_by_uuid_lo(uint64_t uuid_lo);
+
+/*
+ * Persistent memory object
+ */
+
+/*
+ * Object handle
+ */
+typedef struct pmemoid {
+	uint64_t pool_uuid_lo;
+	uint64_t off;
+} PMEMoid;
+
+#define	OID_TYPE(type)							\
+union {									\
+	type *__type;							\
+	PMEMoid oid;							\
+}
+
+#define	OID_ASSIGN(o, value)						\
+	((o).oid = value)
+
+#define	TYPED_OID_ASSIGN(lhs, rhs) (					\
+{(lhs).__type = (rhs).__type; (lhs).oid = (rhs).oid; })
```

Comment:
This may cause problem if it's used in the following way:
```C
OID_TYPE(struct s) x[10];
OID_TYPE(struct s) y[10];
while (i < 10)
    TYPED_OID_ASSIGN(x[i], y[i++]);
```
as `rhs` will be evaluated twice!

What about this:

```C
#define	TYPED_OID_ASSIGN(lhs, rhs) ({ \
    typeof (lhs) _l; \
    typeof (rhs) _r; \
    _l.__type = _r.__type; \
    (lhs).oid = (rhs).oid); })
```